Add options for specifying a minimum workload version requirement

mattleibow opened the pull request:
### Description of Change

If we detect that you have installed .NET MAUI NuGet packages into an SDK which has some old version of the .NET MAUI workload, then we will fail the build.

The error code is MA003 and there is a property to skip this warning as well: `$(SkipCheckForMauiMinimumSupportedWorkloadVersion)`

Errors will look like:

> error MA003: The current .NET MAUI package version 'NUGET_VERSION' requires the .NET MAUI workload version to be at least 'MIN_WORKLOAD_VERSION'. The current .NET MAUI workload version is 'CURRENT_WORKLOAD_VERSION'.
>
> error MA003: The current .NET MAUI package version '9.0.10-dev' requires the .NET MAUI workload version to be at least '9.0.0'. The current .NET MAUI workload version is '8.0.83'.



<!-- Enter description of the fix in this section -->

### Issues Fixed

<!-- Please make sure that there is a bug logged for the issue being fixed. The bug should describe the problem and how to reproduce it. -->

Fixes #25970

<!--
Are you targeting main? All PRs should target the main branch unless otherwise noted.
-->

mattleibow committed:
Add a MSBuild for bad installs

If we detect that you have installed .NET MAUI into a TFM which is not
found in the package, then we will fail the build.

The error code is MA003 and there is a property to skip this warning as
well:

Errors will look like:
- error MA003: This version of .NET MAUI requires at least a
  TargetFramework of 'net9.0-windows10.0.19041.0', which is greater
  than the current TargetFramework 'net8.0-windows10.0.19041.0'.
- error MA003: This version of .NET MAUI requires at least a
  TargetFramework of 'net9.0-ios', which is greater than the current
  TargetFramework 'net8.0-ios'.
- error MA003: This version of .NET MAUI requires at least a
  TargetFramework of 'net9.0-maccatalyst17.0', which is greater than
  the current TargetFramework 'net8.0-maccatalyst17.0'.

jonathanpeppers reviewed (commented):
Maybe @dsplaisted can comment on the general problem here? Targeting .NET 9 MAUI packages from a .NET 8 project would certainly default to `netstandard2.0` if those exist.

@mattleibow why does `Microsoft.Maui.Controls` have `netstandard2.0` assemblies inside _at all_? That is needed for MSBuild tasks inside Visual Studio, so you could simply have a copy of those in the `Microsoft.Maui.Controls.Build.Tasks` package? They shouldn't be referenced by customer projects, just used by the MSBuild tasks?

rolfbjarne commented:
Wild idea: provide a net6.0 version of the assemblies that has:

```cs
[Obsolete ("This version of MAUI requires at least a TargetFramework of ...", IsError = true)]
```

on every API.

mattleibow committed:
Workload versions are better

mattleibow commented on src/Core/src/nuget/buildTransitive/netstandard2.0/Microsoft.Maui.Core.targets:
```diff
@@ -0,0 +1,37 @@
+
+    <Error
+      Code="MA003"
+      Text="This version of .NET MAUI requires at least a TargetFramework of '$(_MauiMinimumTargetFrameworkString)', which is greater than the current TargetFramework '$(TargetFramework)'." />
```
Not sure we have that infrastructure. @tj-devel709 what can we localize?

jonathanpeppers replied:
We should at least create a `<MauiError/>` task and start putting these strings in `.resx` files. We have examples in the dotnet/android repo.

mattleibow replied:
OK, I'll have a look. But, that we can do in a separate PR and I can go through the targets and collect the errors, warnings and messages.

mattleibow replied:
Created an issue https://github.com/dotnet/maui/issues/26308

mattleibow committed:
$(MauiWorkloadVersion) may not be defined

This happens mostly in the maui repo, but also may happen if the
build is not using workloads. This is not really possible outside
of the maui repo source build.

mattleibow committed:
props

jonathanpeppers commented:
> @mattleibow why does Microsoft.Maui.Controls have netstandard2.0 assemblies inside at all? 

Do we know the answer to this question? MSBuild won't even know how to _find_ these assemblies unless you override `AppDomain.AssemblyResolve`. So, we probably have a copy of the `netstandard2.0` assemblies in the `Microsoft.Maui.Controls.Build.Tasks` package.

Redth commented on src/Core/src/Core.csproj:
```diff
@@ -64,15 +64,18 @@
-    <None Include="nuget\**" Exclude="nuget\**\*.in.*" PackagePath="" Pack="true" />
+    <None Remove="nuget\**" />
+    <None Include="nuget\**" Exclude="nuget\**\*.in.*;nuget\**\net-windows\*" PackagePath="" Pack="true" />
+    <None Include="nuget\buildTransitive\net-windows\**" PackagePath="buildTransitive\net$(_MauiMinimumSupportedDotNetTfm)-windows$(MinimumWindowsTargetFrameworkVersion)" Pack="true" />
```
Can you add more context on what this change is doing?

rmarinho replied:
Seems its pick the base targets and moving to correct tfm from for picking the correct windows tfm ?

Redth commented on src/Maui.InTree.props:
```diff
@@ -3,7 +3,7 @@
 
     <!-- Microsoft.Maui.Core.props -->
-    <Import Project="$(_MauiBuildTasksLocation)net6.0-windows10.0.17763.0\Microsoft.Maui.Core.props" Condition="'$(_MauiTargetPlatformIsWindows)' == 'true'" />
+    <Import Project="$(_MauiBuildTasksLocation)net$(_MauiMinimumSupportedDotNetTfm)-windows$(MinimumWindowsTargetFrameworkVersion)\Microsoft.Maui.Core.props" Condition="'$(_MauiTargetPlatformIsWindows)' == 'true'" />
```
Ahh so we used to just keep this very low, now we're setting it to what we know to be the min supported version

rmarinho commented on src/Core/src/nuget/buildTransitive/Microsoft.Maui.Core.After.targets:
```diff
@@ -7,6 +7,32 @@
+     - Installing a .NET 9 MAUI package into a .NET 8 project
+     - Installing a .NET 9 MAUI package into a very old .NET 9 installation
+
+    To skip this check, set $(SkipCheckForMauiMinimumSupportedWorkloadVersion) to true.
```
This will need some documentation heads up @davidortinau  .

mattleibow replied:
Do we _want_ to advertise this option? It will only cause pain, and I just added this as a failsafe just in case.

rmarinho reviewed (changes requested):
Should we add a test where we try to install this on a machine with net8 sdk only? I know it might be tricky. but I think that will cover this edge case works. 

Also I think the message needs to be localized.

rmarinho commented on src/Core/src/nuget/buildTransitive/Microsoft.Maui.Core.After.targets:
```diff
@@ -7,6 +7,32 @@
+    
+    <Error
+        Code="MA003"
+        Text="The current .NET MAUI package version '$(MauiCorePackageVersion)' requires the .NET MAUI workload version to be at least '$(MauiMinimumSupportedWorkloadVersion)'. The current .NET MAUI workload version is '$(MauiWorkloadVersion)'." />
```
I think we need to localize this, now that we have the one loc working back again we might want to create some locale setup for our targets.

rmarinho replied:
Maybe we can create a new issue to follow up on this with @tj-devel709

mattleibow replied:
Created an issue: https://github.com/dotnet/maui/issues/26308

mattleibow committed:
Add a test

PureWeen commented:
/rebase
